### PR TITLE
Add pragmas to hide additional gcc_warnings related to maybe-unitialized memory in std::optionals

### PR DESCRIPTION
**Pull request by jleibs**
### What
We have a handful of places where we need to disable `-Wmaybe-uninitialized` primarily related to usage of `std::optional`.

It seems that move-assigning to the std::optional can lead to situations where gcc thinks we are using uninitialized-memory because the memory in the optional hasn't been initialized.

Tracking these now in: https://github.com/rerun-io/rerun/issues/4027

A few of these still make me uncomfortable, but seems better than having a pile of unusable warning spam.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/4030) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4030)
- [Docs preview](https://rerun.io/preview/f7c01adf1874fed69c13f904d1c810da91e69b0a/docs) 
- [Examples preview](https://rerun.io/preview/f7c01adf1874fed69c13f904d1c810da91e69b0a/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: rerun_cpp/src/rerun/components/rotation3d_ext.cpp [hunk 1]
```diff
@@ -29,6 +29,6 @@ namespace rerun {
 #define Rotation3DExt Rotation3D
 #endif
 
-        const Rotation3D Rotation3D::IDENTITY = datatypes::Quaternion::IDENTITY;
+        const Rotation3DExt Rotation3DExt::IDENTITY = datatypes::Quaternion::IDENTITY;
```
Comment (jleibs) [inline rerun_cpp/src/rerun/components/rotation3d_ext.cpp:32]: I hadn't noticed before we had the define Rotation3DExt Rotation3D above. I think this is actually the right pattern.

File: rerun_cpp/src/rerun/archetypes/depth_image.hpp [hunk 2]
```diff
@@ -110,15 +110,31 @@ namespace rerun {
...
                 draw_order = std::move(_draw_order);
+// See: https://github.com/rerun-io/rerun/issues/4027
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
                 return std::move(*this);
+#ifdef __GNUC__
+#pragma GCC diagnostic pop
+#endif
```
Comment (emilk) [inline rerun_cpp/src/rerun/archetypes/depth_image.hpp:137]: Alternative: we add a `rerun::move` to `rerun/util.hpp` that we call everywhere, making the generated code simpler (as well as the code generator).

But this is fine

File: rerun_cpp/src/rerun/component_batch.hpp [hunk 3]
```diff
@@ -156,7 +155,19 @@ namespace rerun {
...
         void operator=(ComponentBatch<TComponent>&& other) {
+// Need to disable the maybe-uninitialized here.  It seems like the compiler may be confused in situations where
+// we are assigning into an unused optional from a temporary. The fact that this hits the move-assignment without
+// having called the move constructor is suspicious though and hints of an actual bug.
+//
+// See: https://github.com/rerun-io/rerun/issues/4027
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
```
Comment (emilk) [inline rerun_cpp/src/rerun/component_batch.hpp:166]: or maybe make reusable macros for this, i.e.

```C
#define RERUN_PUSH_DIAGNOSTICS_OFF                            \
    #ifdef __GNUC__                                           \
    #  pragma GCC diagnostic push                             \
    #  pragma GCC diagnostic ignored "-Wmaybe-uninitialized"  \
    #endif

#define RERUN_POP_DIAGNOSTICS_TWEAK \
    #ifdef __GNUC__                 \
    #  pragma GCC diagnostic pop  \
    #endif
#endif
```

that we can then call where we need it. This is just so noisy

**Review by emilk — APPROVED** [review]
That was more places than I thought 😬 
I suggest we try to make it a bit more ergonomic by putting this into reusable macros

**Pull request merged**